### ISSUE TITLE
Use kubernetes.io/ingress.class annotation to be able to use with other controllers

### ISSUE THREAD

**Issue by ese** [zalando-incubator/kube-ingress-aws-controller#38]
If the ingress resource has the annotation only take it in account if its match with the controller class. For example `kubernetes.io/ingress.class: alb`

**Comment by hjacobs** [comment]
I would definitely make this configurable (i.e. only require the annotation if it was enabled via command line flag): as a user, I don't want to add an unnecessary annotation if I'm only using one Ingress controller in the whole cluster.

**Comment by mikkeloscar** [comment]
@hjacobs as I understand the issue is to only respect the annotation if it's there, if it's not there the ingress controller just takes full ownership.

Still this could be enabled/disabled with a flag.

**Comment by hjacobs** [comment]
@mikkeloscar right, after reading @ese's text again I understand it in the same way :smile:

**Comment by linki** [comment]
AppsCode's [Voyager](https://github.com/appscode/voyager) has a good approach. Like Mikkel describes and also configurable.

https://github.com/appscode/voyager#running-voyager-alongside-with-other-ingress-controller

**Comment by lmineiro** [comment]
What value should we adopt for this controller? `alb` like suggested?

**Comment by hjacobs** [comment]
We have one potential problem with the annotation: both the `kube-ingress-aws-controller` and Skipper need to listen on changes (in our Zalando setup). Skipper might just read ALL Ingress rules (ignoring the annotation), but otherwise we need to find a way of having multiple values for the annotation (to enable/disable `kube-ingress-aws-controller` and `skipper` separately).

**Comment by szuecs** [comment]
@hjacobs Is it a problem if skipper serves also traffic configured by other ingress-controllers?
I think not, so skipper could just create all rules and if it does not get traffic for one rule, because someone use an ingress-controller that routes traffic to f.e. nginx, there is no problem.

**Comment by lmineiro** [comment]
What @szuecs just mentioned is an important point that I've also been insisting.
The controller deliver things that are pretty standard - A LB with all the worker nodes as members and a DNS name. This can be used as a contract for what it delivers.
I'd even encourage other competing solutions to pick up that contract and transport traffic to applications inside a Kubernetes cluster.
We rely on Skipper and I'm confident this is the right thing for us but an explicit combination of the controller and Skipper should not be required.

**Comment by jhohertz** [comment]
Just ran into this myself. In my case I was attempting to use this + skipper as the main ingress controller, and an instance of ingress-nginx for narrow cases where it was better positioned to handle that traffic.

Took me a while to figure out what was going on, but despite me clearly labelling all my ingress manifests with the annotation kubernetes.io/ingress.class, this component was not respecting this, and I can see it fight with ingress-nginx for those ingress resources in both their logs as it flips back and forth.

As this component is clearly mostly geared towards use with skipper as an ingress router, I would propose that:

1) There be a flag that can be specified on the container definition, IE: --ingress-classes=<comma seperated list of class names
2) This flag defaults to 'skipper' to reflect the most common use-case
3) The controller will avoid causing any action for any class not matching the list in question.
4) Perhaps also offer a wildcard option to preserve existing behavior (IE: for those ONLY using skipper, and not wanting to add the annotations to their manifests)
5) Maybe #4 is the default behaviour as it matches current state... vs #2.

But at least with this it would be flexible enough to interoperate with other ingress controllers, where that is simply impossible in the current state.

**Comment by szuecs** [comment]
@jhohertz I am fine with changing the behaviour. I think 1. can be good enough if it's not specified at all we default to the current behaviour and we document the default behaviour.  
I am also fine with changing the default behaviour and document the update path. @mikkeloscar what do you think about these options?


I checked in skipper we use this as default and override by flag `-kubernetes-ingress-class`:

        ingressClassKey               = "kubernetes.io/ingress.class"
        defaultIngressClass           = "skipper"

**Comment by jhohertz** [comment]
@szuecs That sounds great, and a little more elegant than my suggestion on line 4 regarding that.

Happy to test any efforts toward this. (And I would offer to create a PR, but just getting started w/ Go here...)

**Comment by mikkeloscar** [comment]
@szuecs @jhohertz Keeping the default behavior if you don't specify the flag also sounds like the best approach to me.

I don't want to tie it to skipper by default.

**Comment by hjacobs** [comment]
Who would implement this? Any volunteers?

**Comment by jhohertz** [comment]
I decided to try my hand at this anyways, doesn't look very complicated code-wise, but running into issues w/ build tooling... looking for dependencies in the wrong place. (And my inexperience w/ Go shines...)

I can run glide install w/o error, but run into this running make: ```worker.go:16:2: cannot find package "github.com/aws/aws-sdk-go/aws/awserr" in any of:
	/usr/local/go/src/github.com/aws/aws-sdk-go/aws/awserr (from $GOROOT)
	/home/jhohertz/go/src/github.com/aws/aws-sdk-go/aws/awserr (from $GOPATH)
{clipped more of the same}
```
When I think it should be looking in the vendor folder where glide put the dependencies?

Can anyone clue me in on what I'm doing wrong?

**Comment by szuecs** [comment]
I checked the dependencies and these are specified correctly and I have this github.com/aws/aws-sdk-go/aws/awserr folder in my vendor/ folder of the project.
If you have Go version >1.6, I think it should just work. Glide will download all dependencies to the vendor folder. 

Do you have a vendor folder with the packages that could not be found?
Do you have a recent Go version? (I have go1.10)
Do you have a recent glide version? ( I have glide version 0.13.2-dev)

**Comment by jhohertz** [comment]
Think I figured out my issue, I have to have the project IN $GOPATH/src/github.com/zalando-incubator/kube-aws-ingress-controller. (I had it my usual folder of checkouts..)

**Comment by szuecs** [comment]
Yes in go you have a fixed path from your $GOPATH .

**Comment by jhohertz** [comment]
So I have an initial stab at this pushed up for review/comments (see above). I haven't had a chance yet to build/deploy it in-place in a cluster, but wanted to get it visible for feedback.

**Comment by mikkeloscar** [comment]
@jhohertz please open a PR, then it's easier to review. Doesn't have to be fully working before you open the PR, it's ok to open it as "Work In Progress" :)

**Comment by jhohertz** [comment]
PR #158

**Comment by jhohertz** [comment]
For anyone interested in this, I have a test build of the container w/ the version in PR #158 available at this image: viafoura/kube-ingress-aws-controller:v0.6.8-5-g6321660

**Comment by jhohertz** [comment]
Just a note the PR has been merged, and I can confirm the new functionality is working in the mainline :latest container build.

**Comment by szuecs** [comment]
@jhohertz thanks, so much for your work!

**Comment by szuecs** [comment]
was released in v0.6.9

**Comment by jhohertz** [comment]
@szuecs Thanks for your help making this happen!
